entities/interval/backoff.go

```diff
@@ -0,0 +1,81 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package interval
+
+import (
+	"sort"
+	"time"
+)
+
+const maxInterval = 24 * time.Hour
+
+var defaultBackoffs = []time.Duration{
+	time.Duration(0),
+	30 * time.Second,
+	2 * time.Minute,
+	10 * time.Minute,
+	1 * time.Hour,
+	12 * time.Hour,
+}
+
+// BackoffTimer tracks a given range of intervals with increasing duration
+type BackoffTimer struct {
+	backoffLevel int
+	backoffs     []time.Duration
+	lastInterval time.Time
+}
+
+// NewBackoffTimer constructs and returns a *BackoffTimer instance
+// If no backoffs are provided, defaultBackoffs is used. When the
+// final backoff duration is elapsed, a maximum duration of 24 hrs
```

Comment:
NIT: Could we change it to not have a max of 24hrs and instead just use the last one in the list of backoffs ?

Reasoning is that the 24hrs limit is a fixed upper bound. Maybe in the future we want a max backoff of 2 days and for that we'll have to change the default. Instead if we use the last one in the list then it's a bit more flexible.

Reply:
Sure, sounds sensible to me 🙂